std/net/punycode.d

```diff
@@ -0,0 +1,232 @@
+/**
+	Encodes and decodes punycode strings, per
+	$(LINK2 https://www.ietf.org/rfc/rfc3492.txt, RFC3492).
+	Punycode is primarily used for converting URIs or IRIs containing non-ASCII
+	characters into URIs or IRIs that are entirely ASCII, and vice versa.
+	This punycode codec is based upon the original implementation found in
+	$(LINK2 https://www.ietf.org/rfc/rfc3492.txt, RFC3492).
+
+	Authors: Shotaro Yamada (Sinkuu)
+	Date: February 27, 2017
+	License: $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+	Standards: $(LINK2 https://www.ietf.org/rfc/rfc3492.txt, RFC3492)
+	Source: $(PHOBOSSRC std/net/_punycode.d)
+	Version: 1.0.1
+*/
+module punycode;
+
+private import std.ascii : isASCII, isUpper, isLower, isDigit;
+private import std.conv : to;
+private import std.exception : enforce;
+private import std.traits : isSomeString;
+private import std.array : insertInPlace;
+private import std.algorithm.searching : all;
+
+version (unittest)
+{
+	import std.exception : assertThrown, collectExceptionMsg;
+}
+
+private immutable uint base = 36;
+private immutable ubyte initialN = 0x80;
+private immutable uint initialBias = 72;
+private immutable uint tmin = 1;
+private immutable uint tmax = 26;
+private immutable uint damp = 700;
+private immutable uint skew = 38;
+
+/**
+	Converts an UTF string to a Punycode string.
+	Params: str = A UTF-encoded string that should be encoded into Punycode
+	Returns: A punycode-encoded string
+	Throws: PunycodeException if an internal error occured
+*/
+S punyEncode(S)(in S str) @safe pure
+	if (isSomeString!S)
+{
+	import std.functional : not;
+	import std.algorithm.iteration : filter;
+	import std.array : array, appender, Appender;
+	import std.algorithm.sorting : sort;
+
+	static char encodeDigit(uint x)
+	{
+		if (x <= 25) return cast(char)('a' + x);
+		else if (x <= 35) return cast(char)('0' + x - 26);
+		assert(0);
+	}
+
+	dstring dstr = str.to!dstring;
+	auto ret = appender!S;
+	ret ~= dstr.filter!isASCII;
+	assert(ret.data.length <= uint.max);
+	uint handledLength = cast(uint)ret.data.length;
+	immutable uint basicLength = handledLength;
+	if (handledLength > 0) ret ~= '-';
+	if (handledLength == dstr.length) return ret.data;
+	auto ms = (() @trusted => (cast(uint[])(dstr.filter!(not!isASCII).array)).sort!"a < b")();
```

Comment:
AFAIK are `filter`, `array` and `sort` `@safe`. No?
(the @trusted lambdas should be used on the closest scope possible)